File: urbanairship-core/src/main/java/com/urbanairship/actions/MessageCenterAction.java
/* Copyright Airship and Contributors */

package com.urbanairship.actions;

import androidx.annotation.NonNull;

import com.urbanairship.UAirship;
import com.urbanairship.push.PushMessage;
import com.urbanairship.richpush.RichPushInbox;
import com.urbanairship.richpush.RichPushMessage;
import com.urbanairship.util.UAStringUtil;

/**
 * Starts an activity to display either the {@link RichPushInbox} or a {@link RichPushMessage} using
 * either {@link RichPushInbox#startInboxActivity()} or {@link RichPushInbox#startMessageActivity(String)}.
 * <p>
 * Accepted situations: SITUATION_PUSH_OPENED, SITUATION_WEB_VIEW_INVOCATION,
 * SITUATION_MANUAL_INVOCATION, SITUATION_AUTOMATION, and SITUATION_FOREGROUND_NOTIFICATION_ACTION_BUTTON.
 * <p>
 * Accepted argument values: {@code null} to launch the inbox, the specified message ID, or {@code "auto"}
 * to look for the message ID in the {@link ActionArguments#getMetadata()}.
 * <p>
 * Result value: <code>null</code>
 * <p>
 * Default Registration Names: ^mc, open_mc_action
 */
public class MessageCenterAction extends Action {

    /**
     * Default registry name
     */
    @NonNull
    public static final String DEFAULT_REGISTRY_NAME = "open_mc_action";

    /**
     * Default registry short name
     */
    @NonNull
    public static final String DEFAULT_REGISTRY_SHORT_NAME = "^mc";

    /**
     * Overlay message center registry name
     *
     * @deprecated May be removed in a future SDK version. Use "open_mc_action" instead.
     */
    @NonNull
    @Deprecated
    public static final String REGISTRY_NAME_OVERLAY = "open_mc_overlay_action";

    /**
     * Overlay message center registry short name
     *
     * @deprecated May be removed in a future SDK version. Use "^mc" instead.
     */
    @NonNull
    @Deprecated
    public static final String REGISTRY_SHORT_NAME_OVERLAY = "^mco";

    /**
     * Message ID place holder. Will pull the message ID from the push metadata.
     */
    @NonNull
    public static final String MESSAGE_ID_PLACEHOLDER = "auto";

    @Override
    public boolean acceptsArguments(@NonNull ActionArguments arguments) {
        switch (arguments.getSituation()) {
            case SITUATION_PUSH_OPENED:
            case SITUATION_WEB_VIEW_INVOCATION:
            case SITUATION_MANUAL_INVOCATION:
            case SITUATION_FOREGROUND_NOTIFICATION_ACTION_BUTTON:
            case SITUATION_AUTOMATION:
                return true;

            case SITUATION_BACKGROUND_NOTIFICATION_ACTION_BUTTON:
            case SITUATION_PUSH_RECEIVED:
            default:
                return false;
        }
    }

    @NonNull
    @Override
    public ActionResult perform(@NonNull ActionArguments arguments) {

        String messageId = arguments.getValue().getString();

        if (MESSAGE_ID_PLACEHOLDER.equalsIgnoreCase(messageId)) {
            PushMessage pushMessage = arguments.getMetadata().getParcelable(ActionArguments.PUSH_MESSAGE_METADATA);
            if (pushMessage != null && pushMessage.getRichPushMessageId() != null) {
                messageId = pushMessage.getRichPushMessageId();
            } else if (arguments.getMetadata().containsKey(ActionArguments.RICH_PUSH_ID_METADATA)) {
                messageId = arguments.getMetadata().getString(ActionArguments.RICH_PUSH_ID_METADATA);
            } else {
                messageId = null;
            }
        }

        if (UAStringUtil.isEmpty(messageId)) {
            UAirship.shared().getMessageCenter().showMessageCenter();

        } else {
            UAirship.shared().getMessageCenter().showMessageCenter(messageId);
        }

        return ActionResult.newEmptyResult();
    }

    @Override
    public boolean shouldRunOnMainThread() {
        return true;
    }

}

File: urbanairship-message-center-core/src/main/java/com/urbanairship/messagecenter/actions/MessageCenterAction.kt
/* Copyright Airship and Contributors */
package com.urbanairship.messagecenter.actions

import androidx.annotation.RestrictTo
import androidx.annotation.VisibleForTesting
import com.urbanairship.Airship
import com.urbanairship.actions.Action
import com.urbanairship.actions.ActionArguments
import com.urbanairship.actions.ActionResult
import com.urbanairship.messagecenter.Inbox
import com.urbanairship.messagecenter.Message
import com.urbanairship.messagecenter.MessageCenter
import com.urbanairship.messagecenter.actions.MessageCenterAction.Companion.DEFAULT_NAMES
import com.urbanairship.messagecenter.messageCenter
import com.urbanairship.push.PushMessage
import com.urbanairship.util.UAStringUtil

/**
 * Starts an activity to display either the [Inbox] or a [Message] using
 * either [MessageCenter.showMessageCenter] or [MessageCenter.showMessageCenter].
 *
 * **Accepted situations:**
 * - SITUATION_PUSH_OPENED
 * - SITUATION_WEB_VIEW_INVOCATION
 * - SITUATION_MANUAL_INVOCATION
 * - SITUATION_AUTOMATION
 * - SITUATION_FOREGROUND_NOTIFICATION_ACTION_BUTTON.
 *
 * **Accepted argument values:** `null` to launch the inbox,the specified message ID to display,
 * or `"auto"` to look for the message ID in the [ActionArguments.metadata].
 *
 * **Result value:** `null`
 *
 * **Default Registration Names:** [DEFAULT_NAMES]
 */
public open class MessageCenterAction
@VisibleForTesting internal constructor(
    private val messageCenterProvider: () -> MessageCenter
) : Action() {

    public constructor() : this(
        messageCenterProvider = { Airship.messageCenter }
    )

    override fun acceptsArguments(arguments: ActionArguments): Boolean {
        return when (arguments.situation) {
            Situation.PUSH_OPENED,
            Situation.WEB_VIEW_INVOCATION,
            Situation.MANUAL_INVOCATION,
            Situation.AUTOMATION,
            Situation.FOREGROUND_NOTIFICATION_ACTION_BUTTON -> true
            else -> false
        }
    }

    override fun perform(arguments: ActionArguments): ActionResult {
        val messageCenter: MessageCenter = messageCenterProvider()

        var messageId = arguments.value.string
        if (MESSAGE_ID_PLACEHOLDER.equals(messageId, ignoreCase = true)) {
            val pushMessage =
                arguments.metadata.getParcelable<PushMessage>(ActionArguments.PUSH_MESSAGE_METADATA)

            messageId = if (pushMessage != null && pushMessage.richPushMessageId != null) {
                pushMessage.richPushMessageId
            } else if (arguments.metadata.containsKey(ActionArguments.RICH_PUSH_ID_METADATA)) {
                arguments.metadata.getString(ActionArguments.RICH_PUSH_ID_METADATA)
            } else {
                null
            }
        }

        if (UAStringUtil.isEmpty(messageId)) {
            messageCenter.showMessageCenter()
        } else {
            messageCenter.showMessageCenter(messageId)
        }
        return ActionResult.newEmptyResult()
    }

    override fun shouldRunOnMainThread(): Boolean = true

    /** @hide */
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    public companion object {

        /** Default registry names. */
        public val DEFAULT_NAMES: Set<String> = setOf("open_mc_action", "^mc", "open_mc_overlay_action", "^mco")

        /** Message ID place holder. Will pull the message ID from the push metadata. */
        public const val MESSAGE_ID_PLACEHOLDER: String = "auto"
    }
}
